src/main/java/org/ice4j/ice/CandidatePair.java

```diff
@@ -745,11 +743,42 @@ public Socket getSocket()
     /**
      * Returns the <tt>IceSocketWrapper</tt> for this <tt>CandidatePair</tt>.
      * @return  the <tt>IceSocketWrapper</tt> for this <tt>CandidatePair</tt>.
-     * @deprecated use {@link Component#getSocket()} instead.
      */
     @Deprecated
     public IceSocketWrapper getIceSocketWrapper()
     {
-        return getParentComponent().getSocketWrapper();
+        IceSocketWrapper componentSocket
+            = getParentComponent().getSocketWrapper();
+        // If the merging socket is used, all candidate pairs just refer to the
+        // component.
+        if (componentSocket != null)
+        {
+            return getParentComponent().getSocketWrapper();
+        }
+
+        LocalCandidate localCandidate = getLocalCandidate();
+        if (localCandidate == null)
+        {
+            return null;
+        }
+
+        LocalCandidate base = localCandidate.getBase();
+        if (base != null)
+        {
+            localCandidate = base;
+        }
+
+        RemoteCandidate remoteCandidate = getRemoteCandidate();
+        if (remoteCandidate != null)
+        {
+            SocketAddress remoteAddress
+                = remoteCandidate.getTransportAddress();
+            if (remoteAddress != null)
+            {
+                return localCandidate.getCandidateIceSocketWrapper(remoteAddress);
+            }
+        }
+
+        return localCandidate.getCandidateIceSocketWrapper();
```

Comment:
how does it make sense to have a wrapper but with no remote candidate attached?  will it be updated automatically if/once a remote candidate is determined?

Reply:
I'm not quite sure (this is a rollback from a previous version). But I think if you are using the dynamic ports that you just bind to a local port and it's ready for reading regardless of the remote candidate (though you will need the target if you want to send packets).